repository/repo-operations.go

```diff
@@ -442,16 +456,75 @@ func openPullRequest(config *config.GitXargsConfig, repo *github.Repository, bra
 	// Configure pull request options that the GitHub client accepts when making calls to open new pull requests
 	newPR := &github.NewPullRequest{
 		Title:               github.String(titleToUse),
-		Head:                github.String(branch),
+		Head:                github.String(pr.Branch),
 		Base:                github.String(repoDefaultBranch),
 		Body:                github.String(descriptionToUse),
 		MaintainerCanModify: github.Bool(true),
 		Draft:               github.Bool(config.Draft),
 	}
 
 	// Make a pull request via the Github API
-	pr, resp, err := config.GithubClient.PullRequests.Create(context.Background(), *repo.GetOwner().Login, repo.GetName(), newPR)
+	githubPR, resp, err := config.GithubClient.PullRequests.Create(context.Background(), *pr.Repo.GetOwner().Login, pr.Repo.GetName(), newPR)
+
+	// the go-github library's CheckResponse method can return two different types of rate limiting error:
+	// 1. AbuseRateLimitError which may contain a Retry-After header whose value we can use to slow down, or
+	// 2. RateLimitError which may contain information about when the rate limit will be removed, that we can also use to slow down
+	// Therefore, we need to use type assertions to test for each type of error response, and accordingly fetch the data it may contain
+	// about how long git-xargs should wait before its next attempt to open a pull request
+	githubErr := github.CheckResponse(resp.Response)
+
+	if githubErr != nil {
+
+		var isRateLimited = false
+
+		// Create a new open pull request struct that we'll eventually send on the PRChan
+		opr := types.OpenPrRequest{
+			Repo:    pr.Repo,
+			Branch:  pr.Branch,
+			Retries: 1,
+		}
+
+		// If this request has been seen before, increment its retries count, taking into account previous iterations
+		//Increment the previous count
+		opr.Retries = (pr.Retries + 1)
+
+		// If GitHub returned an error of type RateLimitError, we can attempt to compute the next time to try the request again
+		// by reading its rate limit information
+		if rateLimitError, ok := githubErr.(*github.RateLimitError); ok {
+			isRateLimited = true
+			retryAfter := time.Until(rateLimitError.Rate.Reset.Time)
+			opr.Delay = retryAfter
+			logger.Infof("git-xargs parsed retryAfter %d from GitHub rate limit error's reset time", retryAfter)
+		}
+
+		// If GitHub returned a Retry-After header, use its value, otherwise use the default
+		if abuseRateLimitError, ok := githubErr.(*github.AbuseRateLimitError); ok {
+			isRateLimited = true
+			if abuseRateLimitError.RetryAfter != nil {
+				if abuseRateLimitError.RetryAfter.Seconds() > 0 {
+					opr.Delay = *abuseRateLimitError.RetryAfter
+				}
+			}
+		}
+
+		if isRateLimited {
+			// If we couldn't determine a more accurate delay from GitHub API response headers, then fall back to our user-configurable default
+			if opr.Delay == 0 {
+				opr.Delay = time.Duration(config.SecondsToSleepWhenRateLimited)
+			}
```

Comment:
So this is saying, if the `Delay` is 0, it wasn't determined by our logic, it is golang's default; so set it to our default.

And is my assumption right, that the `Delay` _only_ matters if `isRateLimited`?

Reply:
Yes, precisely! If `Delay` is still `0` by this point, that means we weren't able to read from any GitHub API response headers the actual amount of time they want us to wait until re-trying our request. Hence, we fall back to our default. 

And yes - all this logic is only relevant when git-xargs determines it's been rate-limited (which is done via the `CheckResponse` call and subsequent logic in that codepath).